### PR TITLE
Rely on data-binding to (re)configure descriptor

### PR DESCRIPTION
**Pull request by ndeloof**
rely on `@DataBoundSetter` annotations for data-binding with the `global.jelly` web-UI.
prefer a nested databound component vs a boolean pseudo-property (`useSMTPAuth`)

as Descriptor is re-configured, and not created from scratch by data-binding, need to reset attributes to default values.

Those changes are required by https://github.com/jenkinsci/configuration-as-code-plugin/issues/2

### CODE REVIEW

File: src/main/resources/hudson/tasks/Mailer/help-authentication_zh_CN.html [hunk 1]
```diff
@@ -1,3 +1,3 @@
-<div>
-  当发送邮件时使用SMTP认证.如果你的环境需要使用SMTP认证,在这里指定其用户名和密码.
+<div>
+  当发送邮件时使用SMTP认证.如果你的环境需要使用SMTP认证,在这里指定其用户名和密码.
```
Comment (jglick) [inline src/main/resources/hudson/tasks/Mailer/help-authentication_zh_CN.html]: Do I dare ask?

File: src/main/java/hudson/tasks/SMTPAuthentication.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+public class SMTPAuthentication extends AbstractDescribableImpl<SMTPAuthentication> {
+
+    private String smtpAuthUsername;
+
+    private Secret smtpAuthPassword;
+
+    @DataBoundConstructor
+    public SMTPAuthentication(String smtpAuthUsername, Secret smtpAuthPassword) {
```
Comment (jglick) [inline src/main/java/hudson/tasks/SMTPAuthentication.java]: The `smtpAuth` prefix on these property names is redundant. Better to switch to just `username` & `password`.

File: src/main/resources/hudson/tasks/SMTPAuthentication/config_de.properties [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+Use\ SMTP\ Authentication=Verwende SMTP Authentifizierung
+User\ Name=Benutzername
+Password=Kennwort
```
Comment (jglick) [inline src/main/resources/hudson/tasks/SMTPAuthentication/config_de.properties]: I think you neglected to _remove_ these keys from `Mailer/global_*.properties`.

`jenkinsci/jenkins/core/move-l10n.groovy` is a useful tool that could perhaps be converted to a mojo in `org.jvnet.localizer:maven-localizer-plugin`.

Comment (ndeloof) [reply src/main/resources/hudson/tasks/SMTPAuthentication/config_de.properties]: I indeed didn't know about this script (and others).

File: src/main/resources/hudson/tasks/Mailer/help-authentication_zh_CN.html (hunk 1 above)
Comment (ndeloof) [reply src/main/resources/hudson/tasks/Mailer/help-authentication_zh_CN.html]: Not sure. local git tool only reports file being moved, no internal changes.

**Comment by AndreaGiardini** [comment]
@jglick @ndeloof @kohsuke Sorry to bother you again, can somebody review this pull request? :)

**Comment by finarfin** [comment]
This caused issues with the send test e-mail feature. It was failing at form validation due to field names being different. I sent a PR to the source fork: https://github.com/ndeloof/mailer-plugin/pull/1

**Comment by gomesp** [comment]
It seems that the upstream PR has been merged. Does it mean that the example at https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos/mailer now works (or do we still need this PR to be merged as well)?

**Comment by omegas27** [comment]
@jglick @ndeloof @kohsuke Any news on when it's gonna be merged ? I need the possibility to add authentication from configuration as code plugin

**Comment by oleg-nenashev** [comment]
I have no bandwidth to review it this week. Everybody is busy at Jenkins World, I believe

**Comment by darxriggs** [comment]
ping

**Comment by jglick** [comment]
I am afraid I have no memory of this PR. Needs a maintainer.

File: src/main/resources/hudson/tasks/SMTPAuthentication/config_zh_CN.properties [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/SMTPAuthentication/config_zh_CN.properties]: @ndeloof Why does this properties file contain 13 properties whereas the other ones only contain 3?

File: src/main/resources/hudson/tasks/Mailer/global_fr.properties [hunk 5]
```diff
@@ -30,8 +30,5 @@ Default\ user\ e-mail\ suffix=Suffixe par d\u00E9faut des emails des utilisateur
 Sender\ E-mail\ Address=Adresse e-mail de l''''exp\u00E9diteur
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/Mailer/global_fr.properties]: In `l''''exp\u00E9diteur` there are two extra `''` which should be removed.

File: src/main/resources/hudson/tasks/Mailer/global.properties [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# The MIT License
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/Mailer/global.properties]: @ndeloof Why is this empty "global.properties" file added for `Mailer` but no "config.properties" for `SMTPAuthentication`?

File: src/main/resources/hudson/tasks/SMTPAuthentication/config.jelly [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+-->
+
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
+      <f:entry title="${%User Name}" field="username">
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/SMTPAuthentication/config.jelly]: @ndeloof Indentation (lines 27-32) should be corrected.

File: src/main/resources/hudson/tasks/Mailer/global_es.properties [hunk 8]
```diff
@@ -20,18 +20,15 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
-E-mail\ Notification=Notificaci�n por correo electr�nico
+E-mail\ Notification=Notificaci�n por correo electr�nico
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/Mailer/global_es.properties]: @ndeloof Just looking at the diff on GitHub it looks like the encoding got broken. Could you double-check this (this and other properties files)?

File: src/main/resources/hudson/tasks/Mailer/global_da.properties [hunk 9]
```diff
@@ -25,9 +25,7 @@ E-mail\ Notification=E-mail p\u00e5mindelser
 Default\ user\ e-mail\ suffix=Standard e-mail endelse
 Use\ SSL=Brug SSL
 SMTP\ Port=SMTP Port
-Use\ SMTP\ Authentication=Benyt SMTP Authentificering
```
Comment (darxriggs) [inline src/main/resources/hudson/tasks/Mailer/global_da.properties]: @ndeloof Did you miss to remove `Password` in line 23?

Comment (darxriggs) [inline src/main/resources/hudson/tasks/Mailer/global_da.properties]: @ndeloof `"${%Use SMTP Authentication}"` is still used in "Mailer/global.jelly".
Is it correct to move it from "Mailer/global*.properties" to "SMTPAuthentication/config*.properties"?

**Comment by darxriggs** [comment]
@ndeloof ping

**Comment by jglick** [comment]
I do not think @ndeloof is working on this any more. Would fall to whoever is driving `configuration-as-code` these days. CC @ewelinawilkosz

**Comment by darxriggs** [comment]
@alecharp do you still have this on your list as mentioned in the Jira issue?

**Comment by alecharp** [comment]
Yes I am, sorry I didn't take the time before. Will certainly take it in
the next few days to review and or complete.

Le dim. 14 avr. 2019 à 20:51, René Scheibe <notifications@github.com> a
écrit :

**Comment by darxriggs** [comment]
Thank you very much in advance.

**Comment by thangbn** [comment]
any update guys?

**Comment by pietroaragona** [comment]
Any news about this PR? It is very useful when use JCAC plugin
Thanks

**Comment by h1dden-da3m0n** [comment]
Okay, sorry to make this note here but I believe it will help a few people that find this PR based on the JCasC demos, which state setting SMTP doesn't work.
It dose now, or better since the mailer plugin version 1.22 [issue](https://issues.jenkins-ci.org/browse/JENKINS-53467) [pr](https://github.com/jenkinsci/mailer-plugin/pull/42)).
I'll try and get a PR in with the JCasC demos to fix the statement that it doesn't work.

NOTE: I do not claim to be familiar with this PR or what it intends to change/fix so it might still be useful/necessary to fix other issues. I only posted this here since I got redirected here and just by accident found out that setting SMTP already works now and thought other people might have that same "problem".

**Comment by thangbn** [comment]
> Okay, sorry to make this note here but I believe it will help a few people that find this PR based on the JCasC demos, which state setting SMTP doesn't work.
> It dose now, or better since the mailer plugin version 1.22 [issue](https://issues.jenkins-ci.org/browse/JENKINS-53467), [pr](https://github.com/jenkinsci/mailer-plugin/pull/42)).
> I'll try and get a PR in with the JCasC demos to fix the statement that it doesn't work.
> 
> NOTE: I do not claim to be familiar with this PR or what it intends to change/fix so it might still be useful/necessary to fix other issues. I only posted this here since I got redirected here and just by accident found out that setting SMTP already works now and thought other people might have that same "problem".

awesome, looking forward to it.
thanks

**Comment by fcojfernandez** [comment]
Superseded by #61 which intends to land this PR, which seems to be stalled and with no progress. That PR also fixes the test failure in the PR-builder and adds tests to check the compatibility with JCasC.

@alecharp you can close this PR as your convenience.

**Pull request merged**